demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,134 +142,347 @@ typedef struct
-    {
-        /* Set the members of the publish info. */
-        xPublishInfo.payloadLength = ( uint32_t ) sprintf( cBuffer, ggdDEMO_MQTT_MSG_DISCOVERY, ( long unsigned int ) ulMessageCounter ); /*lint !e586 sprintf can be used for specific demo. */
-        xPublishInfo.pPayload = ( const void * ) cBuffer;
+    /* Reduce ulGlobalEntryTimeMs from obtained time so as to always return the
+     * elapsed time in the application. */
+    ulTimeMs = ( uint32_t ) ( ulTimeMs - ulGlobalEntryTimeMs );
 
-        /* Call the MQTT blocking PUBLISH function. */
-        xMqttStatus = IotMqtt_TimedPublish( mqttConnection,
-                                            &xPublishInfo,
-                                            0,
-                                            _maxCommandTimeMs );
+    return ulTimeMs;
+}
+static BaseType_t prvBackoffForRetry( BackoffAlgorithmContext_t * pxRetryParams )
+{
+    BaseType_t xReturnStatus = pdFAIL;
+    uint16_t usNextRetryBackOff = 0U;
+    BackoffAlgorithmStatus_t xBackoffAlgStatus = BackoffAlgorithmSuccess;
+
+    /**
+     * To calculate the backoff period for the next retry attempt, we will
+     * generate a random number to provide to the backoffAlgorithm library.
+     *
+     * Note: The PKCS11 module is used to generate the random number as it allows access
+     * to a True Random Number Generator (TRNG) if the vendor platform supports it.
+     * It is recommended to use a random number generator seeded with a device-specific
+     * entropy source so that probability of collisions from devices in connection retries
+     * is mitigated.
+     */
+    uint32_t ulRandomNum = 0;
+
+    if( xPkcs11GenerateRandomNumber( ( uint8_t * ) &ulRandomNum,
+                                     sizeof( ulRandomNum ) ) == pdPASS )
+    {
+        /* Get back-off value (in milliseconds) for the next retry attempt. */
+        xBackoffAlgStatus = BackoffAlgorithm_GetNextBackoff( pxRetryParams, ulRandomNum, &usNextRetryBackOff );
 
-        if( xMqttStatus != IOT_MQTT_SUCCESS )
+        if( xBackoffAlgStatus == BackoffAlgorithmRetriesExhausted )
         {
-            IotLogError( "mqtt_client - Failure to publish. Error %s.",
-                         IotMqtt_strerror( xMqttStatus ) );
+            LogError( ( "All retry attempts have exhausted. Operation will not be retried" ) );
         }
+        else if( xBackoffAlgStatus == BackoffAlgorithmSuccess )
+        {
+            /* Perform the backoff delay. */
+            vTaskDelay( pdMS_TO_TICKS( usNextRetryBackOff ) );
 
-        IotClock_SleepMs( _timeBetweenPublishMs );
+            xReturnStatus = pdPASS;
+        }
     }
+    else
+    {
+        LogError( ( "Unable to retry operation with broker: Random number generation failed" ) );
+    }
+
+    return xReturnStatus;
 }
 
-/*-----------------------------------------------------------*/
+static BaseType_t prvCreateMQTTConnectionWithBroker( GGD_HostAddressData_t * pxHostAddressData,
```

Comment:
Is this parameter only used for logging? Can the log just be moved to whatever functions calls this one?